source/extensions/filters/http/common/compressor/compressor.cc

```diff
@@ -0,0 +1,309 @@
+      isEtagAllowed(headers) && isTransferEncodingAllowed(headers) && !headers.ContentEncoding()) {
+    sanitizeEtagHeader(headers);
+    insertVaryHeader(headers);
+    headers.removeContentLength();
+    headers.insertContentEncoding().value(config_->contentEncoding());
+    config_->stats().compressed_.inc();
+  } else if (!skip_compression_) {
+    skip_compression_ = true;
+    config_->stats().not_compressed_.inc();
+  }
+  return Http::FilterHeadersStatus::Continue;
+}
+
+Http::FilterDataStatus CompressorFilter::encodeData(Buffer::Instance& data, bool end_stream) {
+  if (!skip_compression_) {
+    config_->stats().total_uncompressed_bytes_.add(data.length());
+    compressor_->compress(data, end_stream ? Compressor::State::Finish : Compressor::State::Flush);
+    config_->stats().total_compressed_bytes_.add(data.length());
+  }
+  return Http::FilterDataStatus::Continue;
+}
+
+Http::FilterTrailersStatus CompressorFilter::encodeTrailers(Http::HeaderMap&) {
+  if (!skip_compression_) {
+    Buffer::OwnedImpl empty_buffer;
+    compressor_->compress(empty_buffer, Compressor::State::Finish);
+    config_->stats().total_compressed_bytes_.add(empty_buffer.length());
+    encoder_callbacks_->addEncodedData(empty_buffer, true);
+  }
+  return Http::FilterTrailersStatus::Continue;
+}
+
+bool CompressorFilter::hasCacheControlNoTransform(Http::HeaderMap& headers) const {
+  const Http::HeaderEntry* cache_control = headers.CacheControl();
+  if (cache_control) {
+    return StringUtil::caseFindToken(cache_control->value().getStringView(), ",",
+                                     Http::Headers::get().CacheControlValues.NoTransform);
+  }
+
+  return false;
+}
+
+bool CompressorFilter::isAcceptEncodingAllowed(const Http::HeaderMap& headers) const {
+  const Http::HeaderEntry* accept_encoding = headers.AcceptEncoding();
+  if (!accept_encoding) {
+    config_->stats().no_accept_header_.inc();
+    return false;
+  }
+
+  typedef std::pair<absl::string_view, float> encPair; // pair of {encoding, q_value}
+  std::vector<encPair> pairs;
+
+  std::vector<std::string> allowed_compressors = config_->registeredCompressors();
+
+  for (const auto token : StringUtil::splitToken(accept_encoding->value().getStringView(), ",",
+                                                 false /* keep_empty */)) {
+    encPair pair = std::make_pair(StringUtil::trim(StringUtil::cropRight(token, ";")), 1);
+    const auto params = StringUtil::cropLeft(token, ";");
+    if (params != token) {
+      const auto q_value = StringUtil::cropLeft(params, "=");
+      if (q_value != params &&
+          StringUtil::caseCompare("q", StringUtil::trim(StringUtil::cropRight(params, "=")))) {
+        auto result = absl::SimpleAtof(StringUtil::trim(q_value), &pair.second);
+        if (!result) {
+          continue;
+        }
+      }
+    }
+
+    pairs.push_back(pair);
+
+    // Disallow compressors with q=0
+    if (!pair.second) {
```

Comment:
If I understand correctly, only a pair with `q > 0`  should be pushed. Can we just evaluate `q` before pushing it instead of pushing it and then remove it?

Reply:
`pairs` is what a client expects and `allowed_compressors` is what the server can handle. Consider the cases when
```
Accept-Encoding: gzip;q=0, deflate, *
```
and
```
Accept-Encoding: deflate, *
```
whereas the server has only `gzip` configured. If we just exclude the encodings with `q=0` from `pairs` then upon noticing `*` we don't know if `gzip` is acceptable by the client.

Reply:
This makes sense. Do you think we need to have an inline comment on this?